File: android/app/src/main/kotlin/com/app/deenly/deenly/tajweed/TajweedErrors.kt
package com.rnr.deenfocus.tajweed

/** Stable Flutter MethodChannel error codes (ADR-006). Add-only. Mirrors TajweedErrors.swift. */
object TajweedErrorCode {
    const val FEATURE_DISABLED = "FEATURE_DISABLED"
    const val MIC_PERMISSION_DENIED = "MIC_PERMISSION_DENIED"
    const val MIC_BUSY = "MIC_BUSY"
    const val NOT_RECORDING = "NOT_RECORDING"
    const val ALREADY_RECORDING = "ALREADY_RECORDING"
    const val AUDIO_TOO_SHORT = "AUDIO_TOO_SHORT"
    const val AUDIO_TOO_LONG = "AUDIO_TOO_LONG"
    const val AUDIO_QUALITY_POOR = "AUDIO_QUALITY_POOR"
    const val MODEL_MISSING = "MODEL_MISSING"
    const val MODEL_DOWNLOAD_FAILED = "MODEL_DOWNLOAD_FAILED"
    const val MODEL_LOAD_FAILED = "MODEL_LOAD_FAILED"
    const val INFERENCE_FAILED = "INFERENCE_FAILED"
    const val INFERENCE_CANCELLED = "INFERENCE_CANCELLED"
    const val INTERRUPTED = "INTERRUPTED"
    const val UNSUPPORTED = "UNSUPPORTED"
    const val INVALID_ARGS = "INVALID_ARGS"
}

enum class TajweedRecordingState(val wireValue: String) {
    IDLE("idle"),
    RECORDING("recording"),
    SCORING("scoring"),
    CANCELLING("cancelling"),
}

/** Native-side exception carrying a stable [TajweedErrorCode]; never leaks a raw Kotlin exception to Flutter. */
class TajweedNativeException(
    val code: String,
    override val message: String,
) : Exception(message)

File: ios/Runner/Tajweed/TajweedErrors.swift
import Foundation

/// Stable Flutter MethodChannel error codes (ADR-006). Add-only.
enum TajweedErrorCode {
  static let featureDisabled = "FEATURE_DISABLED"
  static let micPermissionDenied = "MIC_PERMISSION_DENIED"
  static let micBusy = "MIC_BUSY"
  static let notRecording = "NOT_RECORDING"
  static let alreadyRecording = "ALREADY_RECORDING"
  static let audioTooShort = "AUDIO_TOO_SHORT"
  static let audioTooLong = "AUDIO_TOO_LONG"
  static let audioQualityPoor = "AUDIO_QUALITY_POOR"
  static let modelMissing = "MODEL_MISSING"
  static let modelDownloadFailed = "MODEL_DOWNLOAD_FAILED"
  static let modelLoadFailed = "MODEL_LOAD_FAILED"
  static let inferenceFailed = "INFERENCE_FAILED"
  static let inferenceCancelled = "INFERENCE_CANCELLED"
  static let interrupted = "INTERRUPTED"
  static let unsupported = "UNSUPPORTED"
  static let invalidArgs = "INVALID_ARGS"
}

enum TajweedRecordingState: String {
  case idle
  case recording
  case scoring
  case cancelling
}

struct TajweedNativeError: Error {
  let code: String
  let message: String

  init(_ code: String, _ message: String) {
    self.code = code
    self.message = message
  }
}
